File: lua/resistor-color-trio/resistor-color-trio.lua
local color_values = {
  black = 0,
  brown = 1,
  red = 2,
  orange = 3,
  yellow = 4,
  green = 5,
  blue = 6,
  violet = 7,
  grey = 8,
  white = 9
}

local function value(color, position)
  if position <= 2 then
    return color_values[color] * 10 ^ (2 - position)
  else
    return 10 ^ color_values[color]
  end
end

local function to_metric_ohms(value)
  if value < 10 ^ 3 then
    return value, "ohms"
  elseif value < 10 ^ 6 then
    return value / 10 ^ 3, "kiloohms"
  elseif value < 10 ^ 9 then
    return value / 10 ^ 6, "megaohms"
  else
    return value / 10 ^ 9, "gigaohms"
  end
end

return {
  decode = function(c1, c2, c3)
    return to_metric_ohms(
      ((value(c1, 1) + value(c2, 2)) * value(c3, 3))
    )
  end
}

File: javascript/resistor-color-trio/resistor-color-trio.js
//@ts-check
//
// This is only a SKELETON file for the 'Resistor Color Trio' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

/**
 * @typedef {"ohms"|"kiloohms"|"megaohms"|"megaohms"|"gigaohms"} MetricOhms
 */

/**
 * 
 * @param {number} value 
 * @returns {[number, MetricOhms]}
 */
function toMetricOhms(value) {
  const [kilo, mega, giga] = [10 ** 3, 10 ** 6, 10 ** 9];
  if (value < kilo) {
    return [value, "ohms"];
  }
  if (value < mega) {
    return [value / kilo, "kiloohms"];
  }
  if (value < giga) {
    return [value / mega, "megaohms"];
  }
  return [value / giga, "gigaohms"];
}

export class ResistorColorTrio {

  static #colorsValues = Object.freeze([
    "black",
    "brown",
    "red",
    "orange",
    "yellow",
    "green",
    "blue",
    "violet",
    "grey",
    "white"
  ]);

  static #colorValue(color) {
    return ResistorColorTrio
      .#colorsValues
      .indexOf(color);
  }

  #value;
  #unit;

  /**
   * 
   * @param {string[]} trio
   */
  constructor(trio) {
    const [a, b, c] = trio.map(ResistorColorTrio.#colorValue);
    
    if([a, b, c].includes(-1)) {
      throw new Error("/invalid color/");
    }
    
    [this.#value, this.#unit] = toMetricOhms((a * 10 + b) * (10 ** c));
  }

  get label() {
    return `Resistor value: ${this.#value} ${this.#unit}`;
  }
}
